src/Connection/Internal/Http2ConnectionProcessor.php

```diff
@@ -1564,4 +1574,41 @@ private function increaseStreamWindow(Http2Stream $stream): void
             );
         }
     }
+
+    private function createStreamInactivityWatcher(int $streamId, int $timeout, ?int $delay = null): ?string
+    {
+        if ($timeout <= 0) {
+            return null;
+        }
+
+        $watcher = Loop::delay($delay ?? $timeout, function () use ($streamId, $timeout): void {
+            if (!isset($this->streams[$streamId])) {
+                return;
+            }
+
+            $stream = $this->streams[$streamId];
+
+            $now = Loop::now();
+
+            if ($stream->lastActivityAt < $now - $timeout) {
+                $this->writeFrame(
+                    Http2Parser::RST_STREAM,
+                    Http2Parser::NO_FLAG,
+                    $streamId,
+                    \pack("N", Http2Parser::CANCEL)
+                );
+
+                $this->releaseStream($streamId, new TimeoutException('The request timed out due to inactivity'));
+                return;
+            }
+
+            $delay = $timeout - ($now - $stream->lastActivityAt);
+
+            $stream->watcher = $this->createStreamInactivityWatcher($streamId, $timeout, $delay);
```

Comment:
Shouldn't we disable / re-enable the watcher on activity instead?

Reply:
Yeah, that's easier. Was thinking it would happen a lot, but no different than how the read watcher behaves.